### PR TITLE
feat: add withdraw for bundlr

### LINKED ISSUE #46: Withdraw money from Bundlr's wallet
After uploading assets, withdraw the excess money funded using: `withdraw(getBalance() - 5000)` or better yet `withdrawAll` when it's available.

We need to be mindful that if we withdraw after each upload, the `uploadMetadata` operation would need to fund twice, therefore, rendering its current optimisation useless.

### PR DESCRIPTION
**Pull request by stranzhay**
closes #46

### CODE REVIEW

**Comment by stranzhay** [comment]
@lorisleiva  between `BundlrStorageDriver` `upload` and `uploadAll` i don't explicitly see those called anywhere in the app. It looks like `upload` calls `uploadAll` so I think it makes sense to do the withdraw there. What do you think of the `getLoadedBalance` on withdraw ?

**Review by lorisleiva — CHANGES_REQUESTED** [review]
Thanks for that, I've added some comments.

One concern I have here is that withdrawing after every upload messes up with some optimisations on the `uploadMetadata` operation.

This operation "pre-funds" with an estimation of all assets + metadata. Then it uploads twice. One for the assets and one for the metadata.

So if we withdraw after the first upload, we'll have to fund again before the second upload, which invalids the optimisation brought by the "pre-funding" step.

The reason for that "pre-funding" optimisation is that, on the browser, funding requires wallet approval. Therefore, by pre-funding, we reduce the amount of wallet pop-up the users have to go through to mint an NFT which is super valuable to the UX.

One solution to this would be to add a `withdraw = true` parameter to the upload methods but we can't do this because these methods are shared with other storage drivers that have no concept of funding or withdrawing. So we need to keep this Bundlr-specific.

I think the best thing to do would be to have a flag on the `BundlrStorageDriver` class directly like `shouldWithdrawAfterUploading` which defaults to true. Then we can have public methods such as `withdrawAfterUploading` and `dontWithdrawAfterUploading` that toggle that flag.

What do you think?

File: src/drivers/storage/BundlrStorageDriver.ts [hunk 1]
```diff
@@ -142,6 +142,17 @@ export class BundlrStorageDriver extends StorageDriver {
     throw new NotYetImplementedError();
   }
 
+  public async withdraw(): Promise<void> {
+    const bundlr = await this.getBundlr();
+    const balance = await bundlr.getLoadedBalance();
+
+    try {
+      await bundlr.withdrawBalance(balance);
```
[the remark this answers is not in the document]
Comment (stranzhay) [reply src/drivers/storage/BundlrStorageDriver.ts:153]: It makes sense why you suggested that in the issue originally now. Also makes sense why i used to get errors on withdraw all the time  in my personal work lol

File: src/drivers/storage/BundlrStorageDriver.ts [hunk 2]
```diff
@@ -142,6 +142,17 @@ export class BundlrStorageDriver extends StorageDriver {
...
+    } catch (err) {
+      throw new BundlrWithdrawError();
```
[the remark this answers is not in the document]
Comment (stranzhay) [reply src/drivers/storage/BundlrStorageDriver.ts:155]: miss on my part for the draft

**Review by lorisleiva — APPROVED** [review]
@stranzhay I allowed myself to make a few changes so the upload plan uses the new withdrawal logic. Let me know what you think.

**Pull request merged**